File: SampleAppObjC/PPHSDKSampleApp/AuthCompletedViewController.m
//
//  AuthCompletedViewController.m
//  PPHSDKSampleApp
//
//  Created by Patil, Mihir on 3/20/18.
//  Copyright © 2018 Patil, Mihir. All rights reserved.
//

#import "AuthCompletedViewController.h"
#import "CaptureAuthViewController.h"
#import "PaymentViewController.h"
#import "UIButton+CustomButton.h"

@interface AuthCompletedViewController ()
@property (weak, nonatomic) IBOutlet UILabel *successMsg;
@property (weak, nonatomic) IBOutlet UIButton *voidAuthBtn;
@property (weak, nonatomic) IBOutlet UITextView *voidCodeViewer;
@property (weak, nonatomic) IBOutlet UIButton *captureAuthBtn;
@property (weak, nonatomic) IBOutlet UITextView *captureAuthCodeViewer;
@property (nonatomic, assign) BOOL isTip;
@end

@implementation AuthCompletedViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setUpDefaultView];
}

- (IBAction)voidAuthorization:(id)sender {
    [PayPalRetailSDK.transactionManager voidAuthorization:self.authTransactionNumber callback:^(PPRetailError *error) {
        if(error != nil) {
            NSLog(@"Error Code: %@", error.code);
            NSLog(@"Error Code: %@", error.message);
            NSLog(@"Error Code: %@", error.debugId);
            return;
        }
        UIAlertController *alert = [UIAlertController
                                     alertControllerWithTitle:@"Void successful"
                                     message:@"The Auth was successfully voided."
                                     preferredStyle: UIAlertControllerStyleAlert];
        UIAlertAction *okButton = [UIAlertAction
                                   actionWithTitle:@"Ok"
                                   style:UIAlertActionStyleDefault
                                   handler:^(UIAlertAction * _Nonnull action) {
                                       UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
                                       PaymentViewController *paymentViewController = [storyboard instantiateViewControllerWithIdentifier:@"PaymentViewController"];
                                       NSMutableArray *navigationArray = [[NSMutableArray alloc] initWithArray: self.navigationController.viewControllers];
                                       [navigationArray removeLastObject];
                                       [navigationArray setObject:paymentViewController atIndexedSubscript:3];
                                       [[self navigationController] setViewControllers:navigationArray animated:YES];
                                   }];
        [alert addAction:okButton];
        [self presentViewController:alert animated:YES completion:nil];
    }];

}

- (IBAction)captureAuthorization:(id)sender {
    
    UIAlertController * alert = [UIAlertController
                                 alertControllerWithTitle:@"Adding a tip?"
                                 message:@"Are you trying to do a regular capture or add a tip to a previous transaction?"
                                 preferredStyle: UIAlertControllerStyleAlert];
    
    UIAlertAction* captureButton = [UIAlertAction
                                actionWithTitle:@"Capture"
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    self.isTip = false;
                                    [self performSegueWithIdentifier:@"goToCaptureAuthView" sender:self];
                                }];
    
    UIAlertAction* tipButton = [UIAlertAction
                                    actionWithTitle:@"Add Tip"
                                    style:UIAlertActionStyleDefault
                                    handler:^(UIAlertAction * action) {
                                        self.isTip = true;
                                        [self performSegueWithIdentifier:@"goToCaptureAuthView" sender:self];
                                    }];
    
    UIAlertAction* cancelButton = [UIAlertAction
                                    actionWithTitle:@"Cancel"
                                    style:UIAlertActionStyleCancel
                                    handler: nil];
    
    //Add your buttons to alert controller
    [alert addAction:captureButton];
    [alert addAction:tipButton];
    [alert addAction:cancelButton];
    
    [self presentViewController:alert animated:YES completion:nil];
}

-(void)setUpDefaultView{
    self.successMsg.text = [NSString stringWithFormat:@"Your authorization was successful: $%@", self.invoice.total];;
    [self.successMsg sizeToFit];
    self.voidCodeViewer.text = @"[PayPalRetailSDK.transactionManager voidAuthorization:self.authTransactionNumber callback:^(PPRetailError *error) { \n <code to handle success/failure> \n }];";
     self.captureAuthCodeViewer.text = @"[PayPalRetailSDK.transactionManager captureAuthorization:self.authTransactionNumber invoiceId:self.invoice.payPalId totalAmount:amountToCapture gratuityAmount:0 currency:self.invoice.currency callback:^(PPRetailError *error, NSString *captureId) { \n <code to handle success/failure> \n }];";
    
    [CustomButton customizeButton:_voidAuthBtn];
    [CustomButton customizeButton:_captureAuthBtn];
}
- (UINavigationController *)getCurrentNavigationController {
    return self.navigationController;
}



#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"goToCaptureAuthView"]) {
        CaptureAuthViewController *captureAuthViewController = [segue destinationViewController];
        captureAuthViewController.authTransactionNumber = self.authTransactionNumber;
        captureAuthViewController.invoice = self.invoice;
        captureAuthViewController.paymentMethod = self.paymentMethod;
        captureAuthViewController.isTip = self.isTip;
    }
}


@end

File: SampleApp/PPHSDKSampleApp/AuthCompletedViewController.swift
//
//  AuthCompletedViewController.swift
//  PPHSDKSampleApp
//
//  Created by Wright, Cory on 1/8/18.
//  Copyright © 2018 cowright. All rights reserved.
//

import UIKit
import PayPalRetailSDK

class AuthCompletedViewController: UIViewController {

    @IBOutlet weak var successMsg: UILabel!
    @IBOutlet weak var voidAuthBtn: CustomButton!
    @IBOutlet weak var voidCodeViewer: UITextView!
    @IBOutlet weak var captureAuthBtn: CustomButton!
    @IBOutlet weak var captureAuthCodeViewer: UITextView!
    
    var invoice: PPRetailInvoice?
    var authTransactionNumber: String?
    var paymentMethod: PPRetailInvoicePaymentMethod?
    var isTip: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setUpDefaultView()
        successMsg.text = "Your authorization of $\(invoice?.total ?? 0) was successful"
        successMsg.sizeToFit()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    @IBAction func voidAuthorization(_ sender: UIButton) {
        
        PayPalRetailSDK.transactionManager().voidAuthorization(authTransactionNumber) { (error) in
            if let err = error {
                print("Error Code: \(String(describing: err.code))")
                print("Error Message: \(String(describing: err.message))")
                print("Debug ID: \(String(describing: err.debugId))")
                return
            }
            
            print("Void Payment was successful.")
            
            let alertController = UIAlertController(title: "Voided Successfully", message: "The payment was voided successfully. Start Over.", preferredStyle: .alert)
            alertController.addAction(UIAlertAction(title: "Start Over", style: .default, handler: { (action) in
                 self.performSegue(withIdentifier: "goToPaymentsView", sender: sender)
            }))
            self.present(alertController, animated: true, completion: nil)
        }
        
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if (segue.identifier == "goToCaptureAuthView") {
            if let captureAuthViewController = segue.destination as? CaptureAuthViewController {
                captureAuthViewController.authTransactionNumber = authTransactionNumber
                captureAuthViewController.invoice = invoice
                captureAuthViewController.paymentMethod = paymentMethod
                captureAuthViewController.isTip = isTip
            }
        }
    }
    
    @IBAction func captureAuthorization(_ sender: UIButton) {
        let alert = UIAlertController(title: "Adding a tip?", message: "Are you trying to do a regular capture or add a tip to a previous transaction?", preferredStyle: .actionSheet)
        
        alert.addAction(UIAlertAction(title: "Capture", style: .default, handler: {action in
            self.isTip = false
            self.performSegue(withIdentifier: "goToCaptureAuthView", sender: sender)
        }))
        alert.addAction(UIAlertAction(title: "Add Tip", style: .default, handler: {action in
            self.isTip = true
            self.performSegue(withIdentifier: "goToCaptureAuthView", sender: sender)
        }))
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        
        if let popoverController = alert.popoverPresentationController {
            popoverController.sourceView = captureAuthBtn
            popoverController.sourceRect = captureAuthBtn.bounds
        }
        
        self.present(alert, animated: true)
    }
    
    func getCurrentNavigationController() -> UINavigationController! {
        return self.navigationController
    }
    
    private func setUpDefaultView(){
        
        voidCodeViewer.text = "PayPalRetailSDK.transactionManager().voidAuthorization(authTransactionNumber) { (error) in\n" +
            "   <code to handle success/failure>\n" +
        "}"
        captureAuthCodeViewer.text = "PayPalRetailSDK.captureAuthorizedTransaction(authTransactionNumber, invoiceId: invoice.payPalId, totalAmount: amountToCapture, gratuityAmount: 0, currency: invoice.currency) { (error, captureId) in\n" +
            "  <code to handle success/failure>\n" +
        "}"
        self.navigationItem.backBarButtonItem = UIBarButtonItem(title: "", style: .plain, target: nil, action: nil)
    }

}
